sql/core/src/test/scala/org/apache/spark/sql/CollationSQLExpressionsSuite.scala

```diff
@@ -1789,44 +1798,90 @@ class CollationSQLExpressionsSuite
+      val tableName = s"t_${t.collationId}_mode_nested_struct1"
       withTable(tableName) {
         sql(s"CREATE TABLE ${tableName}(i STRUCT<f1: STRUCT<f2: STRING COLLATE " +
           t.collationId + ">, f3: INT>) USING parquet")
         sql(s"INSERT INTO ${tableName} VALUES " + valuesToAdd)
         val query = s"SELECT lower(mode(i).f1.f2) FROM ${tableName}"
-        if(t.collationId == "UTF8_LCASE" ||
-          t.collationId == "unicode_ci" ||
-          t.collationId == "unicode") {
-          // Cannot resolve "mode(i)" due to data type mismatch:
-          // Input to function mode was a complex type with strings collated on non-binary
-          // collations, which is not yet supported.. SQLSTATE: 42K09; line 1 pos 13;
-          val params = Seq(("sqlExpr", "\"mode(i)\""),
-            ("msg", "The input to the function 'mode' " +
-              "was a type of binary-unstable type that is not currently supported by mode."),
-            ("hint", "")).toMap
-          checkError(
-            exception = intercept[AnalysisException] {
-              sql(query)
-            },
-            errorClass = "DATATYPE_MISMATCH.TYPE_CHECK_FAILURE_WITH_HINT",
-            parameters = params,
-            queryContext = Array(
-              ExpectedContext(objectType = "",
-                objectName = "",
-                startIndex = 13,
-                stopIndex = 19,
-                fragment = "mode(i)")
-            )
-          )
-        } else {
-          checkAnswer(sql(query), Row(t.result))
-        }
+        checkAnswer(sql(query), Row(t.result))
       }
     })
   }
 
   test("Support mode for string expression with collated strings in array complex type") {
+    case class ModeTestCase[R](collationId: String, bufferValues: Map[String, Long], result: R)
+    val testCases = Seq(
+      ModeTestCase("utf8_binary", Map("a" -> 3L, "b" -> 2L, "B" -> 2L), "a"),
+      ModeTestCase("UTF8_LCASE", Map("a" -> 3L, "b" -> 2L, "B" -> 2L), "b"),
+      ModeTestCase("unicode", Map("a" -> 3L, "b" -> 2L, "B" -> 2L), "a"),
+      ModeTestCase("unicode_ci", Map("a" -> 3L, "b" -> 2L, "B" -> 2L), "b")
+    )
+    testCases.foreach(t => {
+      val valuesToAdd = t.bufferValues.map { case (elt, numRepeats) =>
+        (0L to numRepeats).map(_ => s"array(named_struct('f2', " +
+          s"collate('$elt', '${t.collationId}'), 'f3', 1))").mkString(",")
+      }.mkString(",")
+
+      val tableName = s"t_${t.collationId}_mode_nested_struct2"
+      withTable(tableName) {
+        sql(s"CREATE TABLE ${tableName}(" +
+          s"i ARRAY< STRUCT<f2: STRING COLLATE ${t.collationId}, f3: INT>>)" +
+          s" USING parquet")
+        sql(s"INSERT INTO ${tableName} VALUES " + valuesToAdd)
+        val query = s"SELECT lower(element_at(mode(i).f2, 1)) FROM ${tableName}"
+        checkAnswer(sql(query), Row(t.result))
+      }
+    })
+  }
+
+  test("Support mode for string expression with collated strings in 3D array type") {
```

Comment:
I would modify this test case to something like: Struct<Array<Array<Struct<Struct<String>>>>>

Reply:
@uros-db

I have implemented a test similar to the one you suggested under the name `Support mode for string expression with collated complex type - Highly nested.`

![image](https://github.com/user-attachments/assets/fc9ed1a4-43d5-461f-adb0-edb92289ddc3)

I like this test because by testing a type where an array contains another array directly (i.e., `array<array<...>>`), we can verify the execution path where `getBufferForArrayType` calls `recursivelyGetBufferForArrayType`.

To ensure thorough testing, I may even want to extend this to a 4D array, to confirm that `recursivelyGetBufferForArrayType` can successfully call itself. Essentially, having tests for 3D (and 4D) arrays ensures more comprehensive _branch coverage_. Without these tests, we would primarily achieve only _line coverage_, missing out on deeper validation of recursive behavior.

Thoughts?

Reply:
This recursive implementation does not seem very readable, have you tried something like I proposed above (logic similar to `RewriteCollationJoin`):

can we just have:
`def getCollationAwareBufferForDataType ...`

which does something like
```
    dataType match {
      // Short-circuit if there is no collation.
      case _ if UnsafeRowUtils.isBinaryStable(child.dataType) => ...
      case st: StringType =>
        getCollationAwareBufferForStringType(...)
      case at: ArrayType =>
        getCollationAwareBufferForArrayType(...)
      case  st : StructType =>
        getCollationAwareBufferForStructType(...)
      // Write a comment to point out that don't support MapType (and possibly some other types)...
      case _ => ...
    }
```

And then we can also have:
```
def getCollationAwareBufferForStringType ...
def getCollationAwareBufferForArrayType ...
def getCollationAwareBufferForStructType ...
```
(these can call `getCollationAwareBufferForDataType`, instead of calling each other, to ensure full recursive support)

can we limit the implementation like this? this should help avoid handling "StructType" in `private def recursivelyGetBufferForArrayType` (and vice versa)

Reply:
Yes, that is taken care of. But the existence of `recursivelyGetBufferForArrayType` , for instance, remains. Let's find a way to remove this redundancy -- The question is, Where should I move the call to groupMapReduce? The difference between, for example, recursivelyGetBufferForArrayType and getBufferForArrayType is that getBufferForArrayType calls recursivelyGetBufferForArrayType within the scope of grouping elements together.   If getBufferForArrayType were to recursively call getBufferForArrayType, that wouldn't work.

Reply:
@uros-db 

How about I replace it with this, calling recursive functions within getCollationAwareBuffer, like so:

```
  private def getCollationAwareBuffer(
      childDataType: DataType,
      buffer: OpenHashMap[AnyRef, Long]): Iterable[(AnyRef, Long)] = {
    def getBuffer(groupingFunction: AnyRef => _): Iterable[(AnyRef, Long)] = {
      buffer.groupMapReduce(t =>
        groupingFunction(t._1))(x => x)((x, y) => (x._1, x._2 + y._2)).values
    }
    childDataType match {
      // Short-circuit if there is no collation.
      case _ if UnsafeRowUtils.isBinaryStable(child.dataType) => buffer
      case c: StringType => getBuffer(k =>
        CollationFactory.getCollationKey(k.asInstanceOf[UTF8String], c.collationId))
      case at: ArrayType => getBuffer(k =>
        recursivelyGetBufferForArrayType(at, k.asInstanceOf[ArrayData]))
      case st: StructType =>
        getBuffer(k => recursivelyGetBufferForStructType(
          k.asInstanceOf[InternalRow].toSeq(st).zip(st.fields)))
      // Not supported: MapType
      case _ => buffer
    }
  }
  ```

Reply:
yes, something like this looks fine - like what I was suggesting: `getCollationAwareBuffer` - one central place to handle all dataTypes, and then several helped methods to handle `ArrayType`, `StructType`, `MapType` - these special handlers for array/struct/map can call `getCollationAwareBuffer`, but they should definitely avoid calling each other

so please give it a shot, fiddle around and try to make this approach work

Reply:
@uros-db done! let me know what you think.